### PR TITLE
[WIP] loggable typeclass for log methods

### PR DESCRIPTION
**Pull request by cornerman**
Just a quick poc for having what we have discussed in #51. This is not final. We probably want string overloads without a typeclass. Furthermore, the `Loggable` type is oddly defined in the macors project, because I need to access it in both projects.

Is there a reason for core and macros being split? We already have one thing where a type could be used (https://github.com/outr/scribe/blob/master/macros/shared/src/main/scala/scribe/Macros.scala#L77) with `.tpe =:= typeOf[scribe.Logger]`. Otherwise, we could introduce another subproject for basic types.

### CODE REVIEW

**Comment by darkfrog26** [comment]
The Macros project exists because you can't use a Macro in the same project as the Macro is defined.

**Comment by cornerman** [comment]
I see, we are using the `formatter` macro [here](https://github.com/outr/scribe/blob/b185c99474ede3a745d22f3492ea38111a275c57/core/shared/src/main/scala/scribe/format/Formatter.scala#L10). For the other macros it works, because the macro-based method is not called. (Just pushed a commit to see how it would look like)

File: core/shared/src/main/scala/scribe/LogRecord.scala [hunk 1]
```diff
@@ -37,13 +37,15 @@ trait LogRecord[M] {
 
 object LogRecord {
   object Stringify {
```
Comment (darkfrog26) [inline core/shared/src/main/scala/scribe/LogRecord.scala]: Stringify can be completely removed now with the addition of `Loggable`.

File: core/shared/src/main/scala/scribe/Loggable.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package scribe
+
+trait Loggable[T] {
+  def apply(value: T): String
+}
+
+object Loggable {
```
Comment (darkfrog26) [inline core/shared/src/main/scala/scribe/Loggable.scala]: Would it make sense to add `ThrowableLoggable` here?

**Comment by darkfrog26** [comment]
I really like what you've done here.  Address the `Throwable` scenario properly and I think we're good.

**Comment by cornerman** [comment]
great and done! If we want to merge, I can squash some commits.

An open question is: would it make a performance/codesize difference if we had overloads for string that do not even use the typeclass?

**Comment by darkfrog26** [comment]
It's worth testing to see what kind of performance impact this has.  You may want to test this branch against master using the `benchmark` module:

`jmh:run -i 3 -wi 3 -f1 -t1`

**Comment by darkfrog26** [comment]
Merged. We can play around with it a bit more and tweak performance if it's a problem.

**Comment by darkfrog26** [comment]
@cornerman thanks for your work on this!

**Comment by cornerman** [comment]
Great, one benchmark run just ran out of space :)

**Comment by darkfrog26** [comment]
Testing myself and from before the PR:

```
[info] Benchmark                               Mode  Cnt      Score      Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    3   1473.802 ±  103.614  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    3  28490.985 ± 1662.721  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    3  20464.945 ±  591.156  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    3    554.680 ±   72.785  ns/op
```

```
[info] Benchmark                               Mode  Cnt      Score      Error  Units
[info] LoggingSpeedBenchmark.withLog4j         avgt    3   1500.135 ±  144.828  ns/op
[info] LoggingSpeedBenchmark.withLog4jTrace    avgt    3  28564.537 ±  772.269  ns/op
[info] LoggingSpeedBenchmark.withScalaLogging  avgt    3  20566.571 ± 3305.342  ns/op
[info] LoggingSpeedBenchmark.withScribe        avgt    3    564.292 ±   86.008  ns/op
```

So, within a margin of error, there doesn't seem to be any perceptible impact on performance.

**Comment by cornerman** [comment]
Nice, thank you for testing!

**Pull request merged**